### PR TITLE
Escape arg segments to prevent arbitrary code execution

### PR DESCRIPTION
**Pull request by caseywebdev**
See the test. Arguments to helpers were being passed in unescaped making arbitrary code execution possible. This PR escapes those arguments.

### CODE REVIEW

**Comment by caseywebdev** [comment]
Tests are failing on unrelated code

**Comment by wbond** [comment]
To be clear, this affects situations where users are allowed to supply their own handlebars templates, correct?

Currently all of the tests check the vulnerability via passing string literals of code as the body of the template. Based on this, I presume your testing has indicated that there is no vulnerability if the template is controlled by the site and the user can only provide data?

If that isn't the case, we should add tests for the second class of vulnerability also.

**Comment by wbond** [comment]
BTW, I've fixed up master so that the Travis tests now pass properly

**Comment by caseywebdev** [comment]
AFAIK the issue only exists with a malicious template author.

**Comment by thomasst** [comment]
@wbond Correct. The only vulnerabilities we found is if the user can edit the templates. Let's get this merged.

File: pybars/_compiler.py [hunk 1]
```diff
@@ -595,9 +595,9 @@ def add_block(self, symbol, arguments, nested, alt_nested):
                 u"    options['inverse'] = lambda this: None\n"
                 ])
         self._result.grow([
-            u"    value = helper = helpers.get(u'%s')\n" % symbol,
+            u"    value = helper = helpers.get(%s)\n" % repr(symbol),
```
Comment (wbond) [inline pybars/_compiler.py]: No tests fail when this is reverted

File: pybars/_compiler.py [hunk 2]
```diff
@@ -595,9 +595,9 @@ def add_block(self, symbol, arguments, nested, alt_nested):
...
             u"    if value is None:\n"
-            u"        value = resolve(context, u'%s')\n" % symbol,
+            u"        value = resolve(context, %s)\n" % repr(symbol),
```
Comment (wbond) [inline pybars/_compiler.py]: No tests fail when this is reverted

File: pybars/_compiler.py [hunk 3]
```diff
@@ -690,9 +690,9 @@ def add_invertedblock(self, symbol, arguments, nested, alt_nested):
                 u"    options['fn'] = lambda this: None\n"
                 ])
         self._result.grow([
-            u"    value = helper = helpers.get(u'%s')\n" % symbol,
+            u"    value = helper = helpers.get(%s)\n" % repr(symbol),
```
Comment (wbond) [inline pybars/_compiler.py]: No tests fail when this is reverted

File: pybars/_compiler.py [hunk 4]
```diff
@@ -690,9 +690,9 @@ def add_invertedblock(self, symbol, arguments, nested, alt_nested):
...
             u"    if value is None:\n"
-            u"        value = resolve(context, u'%s')\n" % symbol,
+            u"        value = resolve(context, %s)\n" % repr(symbol),
```
Comment (wbond) [inline pybars/_compiler.py]: No tests fail when this is reverted

File: pybars/_compiler.py [hunk 5]
```diff
@@ -708,7 +708,7 @@ def add_rawblock(self, symbol, arguments, raw):
             u"    options['partials'] = partials\n"
             u"    options['root'] = root\n"
             u"    options['inverse'] = lambda this: None\n"
-            u"    helper = helpers.get(u'%s')\n" % symbol,
+            u"    helper = helpers.get(%s)\n" % repr(symbol),
```
Comment (wbond) [inline pybars/_compiler.py]: No tests fail when this is reverted

**Comment by wbond** [comment]
By the way, if you rebase on master you should get a clean test suite.

**Comment by wbond** [comment]
Just following up here to see if you've had a chance to review my comments?

**Comment by caseywebdev** [comment]
Added `repr(str_class(...))`.

Looks like just python nightly is failing. Seems weird to test nightly since it's not stable, yes?

Regarding the untested sections, I figured the examples given are enough to prove that `repr` is superior to `'%s'` for inserting a string literal for dynamic code generation.

**Comment by wbond** [comment]
Since this is a security issue, I'd really like to have actual tests covering the changes to ensure the fix is properly working and that we don't regress in the future.

**Comment by caseywebdev** [comment]
I really don't have any more time to devote to traveling down each potential code injection source. I find it an obvious improvement to use `repr(untrusted_string)` here over `'%s' % untrusted_string` for code generation since it ensures a string literal where the latter does not, as proven by the added tests.

**Comment by wbond** [comment]
I honestly don't have time for any of this, but am trying to do this out of good will.

Considering I've had multiple people from your org pushing this issue, can someone take the lead?

@thomasst @philfreo

**Comment by philfreo** [comment]
Hey @wbond -- thanks for your hard work on this project/repo.

At this point we're already running a fork in production because there was a gaping security hole in your published version that could lead to arbitrary server-side code execution (also not the first one: https://github.com/wbond/pybars3/pull/3). This PR fixes that and includes tests. Our team has now moved on to other things, but for the sake of all of your package's other users, my suggestion would be doing whatever you can to quickly release a fix for this vulnerability.

**Comment by wbond** [comment]
Basically what I expected. Rather than an org using this in production on a paying app chipping in on an open source project, give the middle finger to someone just helping out in their spare time.

I'll leave this open so everyone knows @closeio isn't interested in helping out. Please refrain from emailing me when you realize you are using a project in way that wasn't intended and are subjecting your users and servers to security issues.

**Comment by wbond** [comment]
Just to be clear, the blocking issue is that the tests don't actually cover what was changed. I can revert more than half of this PR and no tests fail.

Anyone who wants to pitch in and help is more than welcome. I personally don't have the time to do so.

Until then, don't run templates provided by users through pybars3 (or the original pybars).

File: pybars/_compiler.py (hunk 1 above)
Comment (zaafar) [reply pybars/_compiler.py]: @wbond I tried a lot of test-cases to break the following code but failed.

`helpers.get(u'%s')\n" % symbol`

Any symbol I put over here (i.e. **block helper name**) was breaking the template e.g. #foo' #foo; #foo.foo.

Comment (zaafar) [reply pybars/_compiler.py]: So I think this specific change doesn't add any value. Unless I missed something.

**Pull request closed without merging**